File: api/src/main/java/com/avast/metrics/api/Monitor.java
package com.avast.metrics.api;

import java.util.function.Supplier;

public interface Monitor extends AutoCloseable {

    /**
     * Returns a new instance with the given name. This method may be called multiple
     * times - the resulting monitor will remember all its names (basically creating
     * a hierarchy of names).
     *
     * @param name name of the next sub-level
     * @return new monitor derived from this one
     */
    Monitor named(String name);

    /**
     * Returns a new instance with the given names appended.
     *
     * @param name1       name of the next sub-level
     * @param name2       name of the next sub-sub-level
     * @param restOfNames names of the other sub-levels
     * @return new monitor derived from this one
     */
    Monitor named(String name1, String name2, String... restOfNames);

    String getName();

    Meter newMeter(String name);

    Counter newCounter(String name);

    Timer newTimer(String name);

    <T> Gauge<T> newGauge(String name, Supplier<T> gauge);

    Histogram newHistogram(String name);

    void remove(Metric metric);

    @Override
    void close();

}

File: scala-api/src/main/scala/com/avast/metrics/scalaapi/api/Monitor.scala
package com.avast.metrics.scalaapi.api

trait Monitor {
  def named(name: String): Monitor
  def named(name: String, name2: String, names: String*): Monitor
  def getName: String
  def meter(name: String): Meter
  def counter(name: String): Counter
  def timer(name: String): Timer
  def timerPair(name: String): TimerPair
  def gauge[A](name: String)(gauge: () => A): Gauge[A]
  def gauge[A](name: String, replaceExisting: Boolean)(gauge: () => A): Gauge[A]
  def histogram(name: String): Histogram
}
